controllers/bpfman-agent/common.go

```diff
@@ -633,6 +633,31 @@ func (r *ReconcilerCommon) handleProgDelete(
 	return internal.Unchanged, nil
 }
 
+func (r *ReconcilerCommon) unloadAndDeleteApplicationProgramsList(ctx context.Context, bpfDeletedPrograms *bpfmaniov1alpha1.BpfProgramList) error {
+	for _, bpfProgram := range bpfDeletedPrograms.Items {
+		r.Logger.V(1).Info("Deleting bpfProgram", "Name", bpfProgram.Name)
+		id, err := bpfmanagentinternal.GetID(&bpfProgram)
+		if err != nil {
+			r.Logger.Error(err, "Failed to get bpf program ID")
+			return err
+		}
+		r.Logger.Info("Calling bpfman to unload program on node", "bpfProgram Name", bpfProgram.Name, "Program ID", id)
+		if err := bpfmanagentinternal.UnloadBpfmanProgram(ctx, r.BpfmanClient, *id); err != nil {
+			r.Logger.Error(err, "Failed to unload Program")
+			return err
+		}
+
+		r.removeFinalizer(ctx, &bpfProgram, internal.BpfApplicationControllerFinalizer)
```

Comment:
I know it's inefficient, but the way the current controller is implemented, every time we make a change to a BpfProgram object, we exit.  For example, from handleProgDelete:
```
		if r.removeFinalizer(ctx, &bpfProgram, rec.getFinalizer()) {
			return internal.Updated, nil
		}
```